File: Part 8 - Deep Learning/Section 39 - Artificial Neural Networks (ANN)/Ann_MyProgram.py
#Artificial Neural Networks

#Installing Theano








# Part 1 - Data Preprocessing

#importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Churn_Modelling.csv')
X = dataset.iloc[:, 3: 13].values
y = dataset.iloc[:, 13].values

#Encoding the categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X_1 = LabelEncoder()
X[:, 1] = labelencoder_X_1.fit_transform(X[:, 1])
labelencoder_X_2 = LabelEncoder()
X[:, 2] = labelencoder_X_2.fit_transform(X[:, 2])
onehotencoder = OneHotEncoder(categorical_features = [1]) #Creating Dummy Variables
X = onehotencoder.fit_transform(X).toarray()
X = X[:, 1:] #Excludes first dummy variable to avoid dummy variable trap

# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Feature Scaling #Compulsory for ANN
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

# Part 2 - Making the ANN!

#Importing the keras library
import keras #Uses tensorflow in backend
from keras.models import Sequential
from keras.layers import Dense

#Initializing the ANN
classifier= Sequential()

#Adding the input layer and first hidden layer
classifier.add(Dense(output_dim = 6, init = 'uniform', activation = 'relu', input_dim = 11 )) 
#output_dim represents the no of hidden layers (which is generally the average of the input layers and output layers [(11+1)/2)] here, o/p layer is taken as 1 as the dataset has a binary dependent variable
#Adding the second hidden layer
classifier.add(Dense(output_dim = 6, init = 'uniform', activation = 'relu')) #input_dim need not be specified as the hidden layer knows what input node to expect 

#Adding the output layer
classifier.add(Dense(output_dim = 1, init = 'uniform', activation = 'sigmoid'))

#Compiling ANN
classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])
#adam is a type of stochastic gradient descent algorithm
#Fitting ANN to the training set
classifier.fit(X_train, y_train, batch_size = 10, nb_epoch = 100  )
#nb_epoch represents how many times the rounds need to be compiled

# Part - 3 - making the predictions and evaluating the model

y_pred = classifier.predict(X_test)
y_pred = (y_pred > 0.5) #to convert y_pred from probability to True-False values

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

File: Part 8 - Deep Learning/Section 39 - Artificial Neural Networks (ANN)/ann_MyProgram.R
#ANN



# Importing the dataset
dataset = read.csv('Churn_Modelling.csv')
dataset = dataset[4:14]

#Encoding categorical variables as factors
dataset$Geography = as.numeric(factor(dataset$Geography,
                                      levels = c('France', 'Spain', 'Germany'),
                                      labels = c(1, 2, 3)))
dataset$Gender = as.numeric(factor(dataset$Gender,
                         levels = c('Female', 'Male'),
                         labels = c(1, 2)))

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Exited, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Feature Scaling
training_set[-11] = scale(training_set[-11])
test_set[-11] = scale(test_set[-11])

# Fitting ANN to the Training set
install.packages('h2o')
library(h2o)
h2o.init(nthreads = -1) #Allows us to connect to a specific server #nthreads allow us to specify the no of cores to use (-1 selects an optimal value)
classifier = h2o.deeplearning(y = 'Exited',  #To specify the dependent variable
                              training_frame = as.h2o(training_set), #converts the dataset to a h2o frameset
                              activation = 'Rectifier',
                              hidden = c(6,6), #c(6,6) represents no of layers in 1st and 2nd hidden layer
                              epochs = 100, #No of times the ANN algo runs over the dataset 
                              train_samples_per_iteration = -2) #-2 indicates a suitable batch size

# Predicting the Test set results
prob_pred = h2o.predict(classifier, newdata = as.h2o(test_set[-11]))
y_pred = (prob_pred > 0.5)
y_pred = as.vector(y_pred)

# Making the Confusion Matrix
cm = table(test_set[, 11], y_pred)

